src/Chapter3.hs

```diff
@@ -480,6 +520,67 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Building
+  = Church
+  | Library deriving (Show)
+
+data HousePeople
+  = P1
+  | P2
+  | P3
+  | P4 deriving (Show)
+
+peopleCount :: HousePeople -> Int
+peopleCount P1 = 1
+peopleCount P2 = 2
+peopleCount P3 = 3
+peopleCount P4 = 4
+
+newtype House
+  = House { housePeople :: HousePeople }
+  deriving Show
+
+data Wall = Wall deriving (Show)
+
+data City = City
+  { cityCastleName   :: String
+  , cityWalls        :: [Wall]
+  , cityBuildingType :: Building
+  , cityHouses       :: [House]
+  } deriving (Show)
+
+buildCity :: Building -> [House] -> String -> City
+buildCity bulidingType houses "" = City
+  { cityBuildingType = bulidingType
+  , cityHouses       = houses
+  , cityWalls        = []
+  , cityCastleName   = ""
+  }
+buildCity bulidingType houses castleName = City
+  { cityBuildingType = bulidingType
+  , cityHouses       = houses
+  , cityWalls        = [Wall]
+  , cityCastleName   = castleName
+  }
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city { cityCastleName = castleName }
+
+buildHouse :: City -> House -> City
+buildHouse city house = city { cityHouses = house:cityHouses city }
+
+buildWalls :: City -> City
+buildWalls city = city { cityWalls = Wall:cityWalls city }
```

Comment:
This function should build a wall only if all criteria are satisfied. But from the implementation, I see that it always builds walls unconditionally. I see that you've implemented separate helper functions, which is nice :+1: Looks like you can reuse them here :slightly_smiling_face: